### PR TITLE
new demo site with integrated capture and graph/download, plus Cypress tests

### PR DESCRIPTION
**Pull request by TonyIversonLin**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- see README.md for how to run them
* [ ] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [ ] pull request are descriptively named
* [ ] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

File: examples/new-capture/index_copy.html [hunk 1]
```diff
@@ -0,0 +1,363 @@
+<!DOCTYPE html>
```
Comment (jywarren) [inline examples/new-capture/index_copy.html:1]: Just wondering how this is distinct from `index.html`? Thank you!

File: examples/new-capture/index.html [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+  <script src="../../node_modules/flot/jquery.flot.crosshair.js"></script>
+  <script src="../../node_modules/flot/jquery.flot.threshold.js"></script>
+  <script src="../../node_modules/bootstrap/dist/js/bootstrap.min.js"></script>
+  <script src="../../bower_components/jquery.steps/build/jquery.steps.min.js"></script>
+  <script src="graph.js" type="text/javascript"></script>
+  <script src="https://cdn.jsdelivr.net/npm/bs-stepper/dist/js/bs-stepper.min.js"></script>
+
+
+  <script src="capture.js" type="text/javascript"></script>
+  <script src="../../dist/capture.dist.js" type="text/javascript"></script>
```
Comment (jywarren) [inline examples/new-capture/index.html:32]: This is pointed at the compiled file, see review note about Grunt compilation! Thanks!

**Review by jywarren — CHANGES_REQUESTED** [review]
This looks fantastic! A couple small notes and requests:

1. I noticed a `.DS_Store` file was added, probably inadvertently - can you just remove that? Thanks!
2. The copy of the `new-capture` JS files seems like it could diverge development of those files, which may be changing quickly. Was it intentional to make 2 copies of each, and are they exact copies? 
3. The `new-capture` index.html is still pointed at `dist/capture.dist.js`, and that file is compiled from the `examples/capture/*.js` files, not the new `new-capture` copies, so i believe the new copies are not being used in this demo, is that correct? Below is the Gruntfile code which compiles the files:

https://github.com/publiclab/spectral-workbench.js/blob/5b176c52fe9ecf0e4fd533c2390c40b43c783260/Gruntfile.js#L49-L57

Thanks, just need a little clarity on these but overall this is looking really fantastic. I love your final report as well, would you be able to share that in this PR somehow for other contributors to read? 

Great work, we're almost there!! Many thanks!

**Comment by jywarren** [comment]
I'm also seeing an odd error in our Travis build:

```
27.02s$ npm ci 
> cypress@5.1.0 postinstall /home/travis/build/publiclab/spectral-workbench.js/node_modules/cypress
> node index.js --exec install
> cypress@5.1.0 postinstall /home/travis/build/publiclab/spectral-workbench.js/node_modules/cypress
> node index.js --exec install
/home/travis/build/publiclab/spectral-workbench.js/node_modules/cypress/node_modules/fs-extra/lib/mkdirs/make-dir.js:85
      } catch {
              ^
SyntaxError: Unexpected token {
    at createScript (vm.js:80:10)
    at Object.runInThisContext (vm.js:139:10)
    at Module._compile (module.js:617:28)
    at Object.Module._extensions..js (module.js:664:10)
```

https://travis-ci.com/github/publiclab/spectral-workbench.js/builds/182860297#L399-L407

This seems related to the `npm ci` command and/or the `postinstall` stage of cypress. Does this happen to you on local?

**Comment by jywarren** [comment]
And, just confirming that the new capture demo works! Presumably using the compiled (and previous) capture code:

Screen Shot 2020-09-17 at 12 49 06 PM
Screen Shot 2020-09-17 at 12 49 14 PM

A couple icons aren't showing, but it's a minor issue!

**Comment by jywarren** [comment]
Ah, i am getting an error and no graph on the final "save graph" step:

Screen Shot 2020-09-17 at 12 51 13 PM

It looks like `d3` is not getting loaded? 

```
Uncaught ReferenceError: d3 is not defined
    at Class.init (spectral-workbench.js:5178)
    at new Class (spectral-workbench.js:62)
```

`[2020-09-17T16:48:25.081Z]  "GET /bower_components/d3/d3.js" Error (404): "Not found"`

Perhaps it has to be included via NPM instead of bower -- the correct route should be `/node_modules/d3/d3.js`, no?

File: bower.json [hunk 3]
```diff
@@ -2,12 +2,13 @@
...
-    "moment":          "~2.10.6",
-    "nvd3":            "~1.7.1",
-    "d3":              "~3.3.13"
+    "fontawesome": "~4.5.0",
+    "bootstrap-css": "~2.3.2",
+    "jquery": "~2.1.3",
+    "moment": "~2.10.6",
+    "nvd3": "~1.7.1",
+    "d3": "~3.3.13",
+    "jquery.steps": "^1.1.0"
```
Comment (jywarren) [inline bower.json:11]: Here I think we should move this to the `package.json` file, as we are trying to stop using Bower. I believe all existing dependencies have already been moved, too!

File: examples/new-capture/index.html [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bs-stepper/dist/css/bs-stepper.min.css">
+
+  <script src="https://kit.fontawesome.com/a076d05399.js"></script>
+  <script src="../../node_modules/getusermedia-js/dist/getUserMedia.min.js"></script>
+  <script src="../../node_modules/jquery/dist/jquery.min.js"></script>
+  <script src="../../node_modules/flot/jquery.flot.js"></script>
+  <script src="../../node_modules/flot/jquery.flot.crosshair.js"></script>
+  <script src="../../node_modules/flot/jquery.flot.threshold.js"></script>
+  <script src="../../node_modules/bootstrap/dist/js/bootstrap.min.js"></script>
+  <script src="../../bower_components/jquery.steps/build/jquery.steps.min.js"></script>
```
Comment (jywarren) [inline examples/new-capture/index.html:26]: Yes - here, and below, all instances of `bower_components` should be changed to `node_modules` and all dependencies but `jquery.steps` have been transferred there already. That should resolve the last error on the final "plotting" page of the new demo!

File: bower.json (hunk 3 above)
Comment (jywarren) [reply bower.json:11]: https://github.com/publiclab/spectral-workbench.js/blob/5b176c52fe9ecf0e4fd533c2390c40b43c783260/package.json#L27

**Comment by TonyIversonLin** [comment]
If you open the original index.html file, it has a hyper link to the new
demo site. Also please do another npm install and bower install before
running the new demo page. Thanks.

**Comment by TonyIversonLin** [comment]
Awesome

**Comment by TonyIversonLin** [comment]
Hi Jeff,

that is correct. The new demo did not use the new capture module which
allow user to adjust resolution.

**Comment by jywarren** [comment]
Hi @TonyIversonLin - so, can we delete the extra copies of the JS files in the `new-capture` folder, since it's all running from the `/dist/____` versions anyways? Thank you!

**Comment by TonyIversonLin** [comment]
Yes I think we can delete those.

**Comment by jywarren** [comment]
I believe this is ready to go except for the Cypress-related test error!

Noting that we do need a way to download the spectrum output in a file, either CSV or JSON. We can do this in follow-up.

**Comment by jywarren** [comment]
Ohoho -- `The command "bower install" failed and exited with 1 during .`

**Comment by jywarren** [comment]
Hmm, back to:

```
> cypress@5.1.0 postinstall /home/travis/build/publiclab/spectral-workbench.js/node_modules/cypress
> node index.js --exec install
/home/travis/build/publiclab/spectral-workbench.js/node_modules/cypress/node_modules/fs-extra/lib/mkdirs/make-dir.js:85
      } catch {
              ^
```

We should search for that error to see what's up with Cypress!

**Comment by TonyIversonLin** [comment]
Definitely.  We don't really need the cypress test.

**Comment by jywarren** [comment]
Maybe https://on.cypress.io/using-cypress-faq#Why-do-my-Cypress-tests-pass-locally-but-not-in-CI 

Or https://github.com/cypress-io/cypress/issues/1297 ?

File: .travis.yml [hunk 5]
```diff
@@ -3,9 +3,6 @@ node_js:
   - '8'
```
Comment (jywarren) [inline .travis.yml:3]: https://docs.cypress.io/guides/guides/continuous-integration.html#Travis

File: .travis.yml [hunk 6]
```diff
@@ -1,16 +1,26 @@
 language: node_js
```
Comment (jywarren) [inline .travis.yml:1]: And another example here: https://github.com/cypress-io/cypress-example-kitchensink/blob/master/.travis.yml

**Comment by jywarren** [comment]
Almost there!!!!

```
http://127.0.0.1:8080/examples/new-capture/
We attempted to make an http request to this URL but the request failed without a response.
We received this error at the network level:
  > Error: connect ECONNREFUSED 127.0.0.1:8080
```

OK, so we need to point that at the correct page!

File: cypress/integration/spectral-workbench-demo.spec.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+/// <reference types="cypress" />
+
+context('Actions', () => {
+
+  beforeEach(() => {
+    cy.visit('http://127.0.0.1:8080/examples/new-capture/')
```
Comment (jywarren) [inline cypress/integration/spectral-workbench-demo.spec.js:6]: Let's change this!

**Comment by jywarren** [comment]
Yay! The test is running!

```
  1) Actions
       "before each" hook for "It reach the default landing page":
     ReferenceError: The following error originated from your application code, not from Cypress.
  > AdapterJS is not defined
When Cypress detects uncaught errors originating from your application it will automatically fail the current test.
```

**Comment by jywarren** [comment]
Something's wrong with d3...

```
   - can export SVG...X can export SVG
     TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id') in file:///home/travis/build/publiclab/spectral-workbench.js/dist/spectral-workbench.js (line 5374) (1)
```

File: Gruntfile.js [hunk 8]
```diff
@@ -64,10 +64,10 @@ module.exports = function(grunt) {
               specs: 'spec/javascripts/*spec.js',
               vendor: [
                 'node_modules/jquery/dist/jquery.min.js',
-                'bower_components/d3/d3.js',
-                'bower_components/nvd3/build/nv.d3.js',
-                'bower_components/bootstrap-css/js/bootstrap.min.js',
-                'bower_components/moment/moment.js',
+                'node_modules/d3/d3.js',
```
Comment (jywarren) [inline Gruntfile.js:67]: Could something be wrong with this line, for jasmine?

**Comment by jywarren** [comment]
Oh hmm, a bunch of Jasmine is broken:

```

     TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id') in file:///home/travis/build/publiclab/spectral-workbench.js/dist/spectral-workbench.js (line 5374) (1)
   - can copy calibrations with copyCalibration()...X can copy calibrations with copyCalibration()
     TypeError: undefined is not an object (evaluating 'graph.datum') in file:///home/travis/build/publiclab/spectral-workbench.js/spec/javascripts/api_spec.js (line 128) (1)
   - generates copyCalibration callback...X generates copyCalibration callback
     Expected spy success to have been called. (1)
   - can transform graph data...X can transform graph data
     TypeError: undefined is not an object (evaluating 'graph.datum') in file:///home/travis/build/publiclab/spectral-workbench.js/spec/javascripts/api_spec.js (line 156) (1)
   - can limit graphed data to wavelength range...X can limit graphed data to wavelength range
     TypeError: undefined is not an object (evaluating 'graph.datum') in file:///home/travis/build/publiclab/spectral-workbench.js/spec/javascripts/api_spec.js (line 168) (1)
   - can blend graphed spectrum data with that of another spectrum...X can blend graphed spectrum data with that of another spectrum
```

File: Gruntfile.js (hunk 8 above)
Comment (jywarren) [reply Gruntfile.js:67]: Hmm. `  <script src="../../node_modules/d3/d3.js" type="text/javascript"></script>` indicates not...

**Comment by jywarren** [comment]
OK, this is odd. But regarding:

>      TypeError: undefined is not an object (evaluating 'graph.datum') in file:///home/travis/build/publiclab/spectral-workbench.js/spec/javascripts/api_spec.js (line 128) (1)

That `graph` object is defined in the previous test: 

https://github.com/publiclab/spectral-workbench.js/blob/8d0526d9c7efbb80b3356582b40ce3546b6c785a/spec/javascripts/api_spec.js#L103

Is there an order of execution issue happening? No, i think perhaps the SVG test failure is causing the `graph` object not to be initialized. Let's focus on the SVG test.

**Comment by jywarren** [comment]
OK, reproduced it by going to the graph example at /examples/index.html, copying `/spec/javascript/fixtures/spectrum-9.json` to `/spectrums/9.json` (so the aync fetch of the data works), then running:

```js
graph2 = new SpectralWorkbench.Graph({
      spectrum_id: 9,
      calibrated: true,
      onImageComplete: function() { done(); }
    });
```

Here's the full trace:

```
Uncaught TypeError: Cannot read property 'id' of undefined
    at SVGGElement.<anonymous> (spectral-workbench.js:5374)
    at SVGGElement.o (d3.min.js:1)
    at d3.min.js:3
    at Y (d3.min.js:1)
    at Array.Co.each (d3.min.js:3)
    at Array.Co.attr (d3.min.js:3)
    at _graph.load (spectral-workbench.js:5373)
    at Object.success (spectral-workbench.js:127)
    at c (jquery.min.js:2)
    at Object.fireWith [as resolveWith] (jquery.min.js:2)
```

**Comment by jywarren** [comment]
OK - this is a known error: https://github.com/publiclab/spectral-workbench.js/pull/165

Don't know why it's just suddenly showing up on Travis. Let's downgrade again and see.

**Comment by jywarren** [comment]
Trying to bump d3 to `v3.5.4` because of https://github.com/publiclab/spectral-workbench.js/issues/180#issuecomment-701097109

graph demo works:

image

new capture demo works:

image

I think we could do this... but will `nvd3` work?

**Comment by jywarren** [comment]
Aha ok, so: noting `"GET /node_modules/jquery.steps/build/jquery.steps.min.js" Error (404): "Not found"`

and, back to the original error:

```

     TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id') in file:///home/travis/build/publiclab/spectral-workbench.js/dist/spectral-workbench.js (line 5374) (1)
   - can copy calibrations with copyCalibration()...X can copy calibrations with copyCalibration()
     TypeError: undefined is not an object (evaluating 'graph.datum') in file:///home/travis/build/publiclab/spectral-workbench.js/spec/javascripts/api_spec.js (line 128) (1)
```

**Comment by jywarren** [comment]
OK so we have trouble with d3 v3.5 - the known error in our codebase: #165

But v3.4 had caused us trouble with `contextify` -- specifically we bumped to v3.5.4 to resolve contextify issues in
02c1adc

So we are left with: 

1. solve the known error with `TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id') in file:///home/travis/build/publiclab/spectral-workbench.js/dist/spectral-workbench.js (line 5374) (1)`
2. find a pre-`v3.5.4` (or post!) version that does not have that error?

**Comment by jywarren** [comment]
So, noting that this error would occur a LOT in this file:

https://github.com/publiclab/spectral-workbench.js/blob/7b52a773650951ab3a2b54c6ad60f8f117a6d72d/src/SpectralWorkbench.Graph.js#L234

because this segment causing it is repeatedly called in that file:

```js
var id = d3.select('svg').data()[0][index].id; // this is the real d3 DOM-stored data
```

**Comment by jywarren** [comment]
Noting the error occurs in client side of the graph demo!

image

**Comment by jywarren** [comment]
Hmm, maybe the two errors are different, `Cannot read property 'id' of undefined` vs `TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id')` -- but they seem to affect almost the same segment of code.

**Comment by jywarren** [comment]
Solved `TypeError: undefined is not an object (evaluating 'd3.select('svg').data()[0][index].id') in file:///home/travis/build/publiclab/spectral-workbench.js/dist/spectral-workbench.js (line 5374) (1)` !!

Now we have one more error -- all these are returning incorrectly... could be related:

```
   X should initialize line styling and table rows properly
     Expected 1 to be 2. (1)
     Expected 1 to be 2. (2)
     Expected 1 to be 2. (3)
     Expected 1 to be 2. (4)
```

https://github.com/publiclab/spectral-workbench.js/blob/7b52a773650951ab3a2b54c6ad60f8f117a6d72d/spec/javascripts/set_spec.js#L54-L63

**Comment by jywarren** [comment]
No more test failures!

Just:

```
101 specs in 14.061s.
>> 0 failures
Fatal error: Callback must be a function
The command "grunt jasmine" exited with 3.
```

**Comment by jywarren** [comment]
Related to node 10, switched to 12! https://github.com/gruntjs/grunt-contrib-jasmine/issues/266

**Comment by jywarren** [comment]
YES!!!! All done! Merging! Thank you!!!!

**Pull request merged**